perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3327,434 +3338,488 @@ private boolean isThisLightweightSynchronization(PerunSession sess, Group group)
-				continue;
-			} catch (AttributeValueException e) {
-				// There is a problem with attribute value, so set INVALID status of the member
-				getPerunBl().getMembersManagerBl().invalidateMember(sess, member);
 			}
+			log.info("Group synchronization {}: New member id {} added.", group, member.getId());
+		} catch (AlreadyMemberException e) {
+			//This part is ok, it means someone add member before synchronization ends, log it and skip this member
+			log.debug("Member {} was added to group {} before adding process. Skip this member.", member, group);
+			return;
+		} catch (AttributeValueException e) {
+			// There is a problem with attribute value, so set INVALID status of the member
+			getPerunBl().getMembersManagerBl().invalidateMember(sess, member);
+		}
 
-			// Try to validate member
-			try {
-				getPerunBl().getMembersManagerBl().validateMember(sess, member);
-			} catch (AttributeValueException e) {
-				log.warn("Member id {} will be in INVALID status due to wrong attributes {}.", member.getId(), e);
-			}
+		// Try to validate member
+		try {
+			getPerunBl().getMembersManagerBl().validateMember(sess, member);
+		} catch (AttributeValueException e) {
+			log.warn("Member id {} will be in INVALID status due to wrong attributes {}.", member.getId(), e);
 		}
 	}
 
 	/**
-	 * Remove former members from group (if they are not listed in ExtSource yet).
+	 * Remove former member from group (if he is not listed in ExtSource).
 	 *
 	 * If this is membersGroup (of some Vo) try to disableMember, if not possible then delete him.
 	 * If this is regular group (of some Vo) remove him and if this group is also
 	 * his last authoritative group, disable or delete him also in the Vo.
 	 *
-	 * @param sess
+	 * This method runs in separate transaction.
+	 *
+	 * @param sess perun session
 	 * @param group to be synchronized
-	 * @param membersToRemove list of members to be removed from Group
+	 * @param memberToRemove member to be removed from Group
 	 *
-	 * @throws InternalErrorException if some internal error occurs
+	 * @throws WrongAttributeAssignmentException if there is some problem with assignment of attribute
+	 * @throws GroupNotExistsException if group does not exist
+	 */
+	public void removeFormerMemberWhileSynchronization(PerunSession sess, Group group, RichMember memberToRemove) throws WrongAttributeAssignmentException, GroupNotExistsException {
+		// Member is missing in the external group, so remove him from the perun group
+		try {
+			//members group
+			if (group.getName().equals(VosManager.MEMBERS_GROUP)) {
+				// If the group is members group, the member must be disabled as a member of VO
+				removeMemberFromMembersGroup(sess, group, memberToRemove);
+			//not members group
+			} else {
+				//If this group is authoritative group, check if this is last authoritative group of this member
+				//If Yes = deleteMember (from Vo), if No = only removeMember
+				if (isAuthoritative(sess, group)) {
```

Comment:
We could probably resolve ``authoritative`` status of the group outside just once and pass it into this method as parameter. Like this we are getting that attribute n-times (removed members count).

Reply:
Yes, that is better, fixed.